jax/lax/lax.py

```diff
@@ -2451,6 +2451,8 @@ def _convert_element_type_translation_rule(c, operand, *, new_dtype, old_dtype):
 
 def _convert_element_type_transpose_rule(t, *, new_dtype, old_dtype):
   assert t.dtype == new_dtype, (t.dtype, new_dtype)
+  if core.primal_dtype_to_tangent_dtype(old_dtype) == dtypes.float0:
+    return [ad_util.Zero(core.Float0Array(t.shape))]
```

Comment:
This special case was necessary to fix the `test_vjp_of_int_fulllike ` test. Without this fix that test fails with an assert failure here: https://github.com/google/jax/blob/04f9a7e53db3eb0eb6d5d08a5f2a967bf9712048/jax/interpreters/ad.py#L158. This happened in quite a few places because some `lax` primitives use `full_like` with an integer constant.

My working theory is this only happens when tracking the tangent of a ConstVar in the `backward_pass`, and this tangent not matching the cotangent generated by traversing the Jaxpr backwards through a `convert_element_type`.  It only happens when the tangent is a known `PartialValue`, otherwise the `float0` cotangent is wrapped in a `Zero` and correctly ignored. I think (?) the `PartialValue` is only known when the variable we are tracking is a ConstVar, but I'm very unsure there.

It feels like maybe there is a better fix here: should we be tracking the tangent of ConstVars in the first place? Is there a more central place to guard against this "TypeError" than in the transpose rules?

Reply:
IIRC we think this might be an issue with `convert_element_type`'s JVP rule not being correct (i.e. we shouldn't treat it as linear, since e.g. the JVP rule for `lambda x: convert_element_type(x, old_dtype=float32, new_dtype=int32)` should be `lambda x, t: (convert_element_type(x, old_dtype=float32, new_dtype=int32), convert_element_type(t, old_dtype=float32, new_dtype=float0)` (or if we prefer the second element of the return value should be a symbolic zero with float0 as its aval's dtype).

Is that right?

Reply:
yes! have now updated the jvp rule of  `convert_element_type` to return symbolic zeros when `new_dtype` is int/bool and the transpose_rule to do the same for `old_dtype`, I _think_ that's how we discussed it.